### ISSUE TITLE
Subscribing to SECURITY_ARM_STATE events

### ISSUE THREAD

**Issue by mattstrom** [SmartThingsCommunity/smartapp-sdk-nodejs#145]
**What is your question? Please describe.**
I am interesting in listening to `SECURITY_ARM_STATE` events in my SmartApp. In reading through the Node SDK code, I discovered 2 relevant methods: `subscribedSecurityArmStateEventHandler()` and `subscribeToSecuritySystem()`. However, when I attempt to use them, the API throws an 403 error. How are these methods meant to be used?

**Additional context**
When I attempt to use them, the API throws the following error.

```
ClientError {
  name: 'SmartThingsApiError',
  statusCode: 403,
  message: '403 - undefined',
  url: 'https://api.smartthings.com/installedapps/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/subscriptions',
  body: '{"sourceType":"SECURITY_ARM_STATE","securityArmState":{"locationId":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","subscriptionName":"securityArmState"}}'
}
```

The error is obviously a `403 Forbidden` error, but I have permitted all of the scopes for my app. Somewhere I found that the `r:locations:*` permission is necessary. However, I couldn't find a way to grant this permission during the app installation process. Using `.permissions()` with any list of permissions actually makes the installation process bomb.

Code (derived mostly from sample Node app in this repo):
```
const smartApp = new SmartApp()
    .clientId(CLIENT_ID)
    .clientSecret(CLIENT_SECRET)
    .publicKey('@smartthings_rsa.pub')
    .enableEventLogging(2)
    .configureI18n()
    .permissions([
      // Uncommenting any or all of these makes app installation error out
      // 'i:deviceprofiles:*',
      // 'r:locations:*',
      // 'r:devices:*',
      // 'w:devices:*',
      // 'x:devices:*',
      // 'r:scenes:*',
      // 'x:scenes:*',
    ])
    .page('mainPage', (context, page, configData) => {
      page.section('sensors', section => {
        section
          .deviceSetting('contactSensors')
          .capabilities(['contactSensor'])
          .multiple(true)
          .required(false);

        section
          .deviceSetting('motionSensors')
          .capabilities(['motionSensor'])
          .multiple(true)
          .required(false);
      });

      page.section('devicesToControl', section => {
        section
          .deviceSetting('lights')
          .capabilities(['switch'])
          .multiple(true)
          .required(true)
          .permissions('rx');
      });
    })
    .updated(async (context, updateData) => {
      // Called for both INSTALLED and UPDATED lifecycle events if there is no separate installed() handler
      try {
        await context.api.subscriptions.unsubscribeAll();
        await context.api.subscriptions.subscribeToSecuritySystem('securityArmState');

        return context.api.subscriptions.subscribeToDevices(
          context.config.motionSensors,
          'motionSensor',
          'motion',
          'myDeviceEventHandler'
        );
      } catch (err) {
        console.error(err);
      }
    })
    .subscribedEventHandler('myDeviceEventHandler', (context, event) => {
      const value = (event.value === 'active')
        ? 'on'
        : 'off';

      context.api.devices.sendCommands(context.config.lights, 'switch', value);
    })
    .subscribedSecurityArmStateEventHandler('securityArmState', (event) => {
      console.log(event.armState);
    });
```

**Comment by bflorian** [comment]
There is a separate scope required for subscribing to security system events, `r:security:locations:*:armstate`, as described in the [API Reference Documentation](https://smartthings.developer.samsung.com/docs/api-ref/st-api.html#operation/saveSubscription). Unfortunately the Developer workspace is lagging the documentation in that it doesn't allow you to choose this scope when you are defining your app. I'll get to the work-around for that in a minute.

Generally app permissions work in this way:
1. You must include all scopes that your app might need when you define the app in the dev workspace
2. You don't need to set permissions with the `.permissions(...)` call if all your app does is use devices that are selected via the configuration pages. The system knows which devices these are, so you don't need to specify anything in the `.permissions(...)` call, but you do need to specify the scopes when defining the app in the dev workspace. For device subscriptions and control these would be the `r:devices:*` and `x:devices:*` scopes.
3. If you want to do anything other than read and control devices or if you want to be able to read and control devices that aren't selected in configuration pages, then you need to specify the scopes in the permissions method call. For example, if you wanted to be able to list all devices in a location and read the status of any of them, you would make this call `.permissions('r:devices:*'). But if you make that call and you haven't included that scope when defining the app, then you'll get an error.

You can work around the inability to add the security system permission in the dev workspace using a Personal Access Token (PAT) and following these steps:

>- Get a personal access token with at least `w:apps` scope from [https://account.smartthings.com/tokens](https://account.smartthings.com/tokens)
>
>- Save the current OAuth information for you app by running the following command, substituting your PAT_TOKEN and APP_ID. 
>get your APP_ID from the Developer Workspace
> ```$bash
> curl -H "Authorization: Bearer {PAT_TOKEN}" \
>      https://api.smartthings.com/apps/{APP_ID}/oauth > oauth.json
> ```
>
> - Edit the `oauth.json` file to add the alarm system scope. It should look something like this (assuming you also need to control devices in your app):
>```$json
> {
>     "clientName": "Alarm system example",
>     "scope": [
>       "r:devices:*",
>       "x:devices:*",
>       "r:security:locations:*:armstate",
>     ],
>     "redirectUris": []
>   }
>```
>
> - Update your app with the new OAuth scopes:
>```$bash
> curl -X PUT -H "Authorization: Bearer {PAT_TOKEN}" \
>      -d @oauth.json \
>      https://api.smartthings.com/apps/{APP_ID}/oauth
>```
>
We have created a ticket to add the missing scopes to the dev workspace.

**Comment by mattstrom** [comment]
Thank you for the workaround. However, I don't think it was sufficient.

I made a PUT request to update the scopes for my app. Making a GET request afterwards yields this payload:
```
{
    "clientName": "Alarm system example",
    "scope": [
        "i:deviceprofiles:$",
        "r:locations:*",
        "r:devices:$",
        "r:scenes:*",
        "x:devices:*",
        "r:security:locations:*:armstate",
        "w:devices:$",
        "r:devices:*",
        "w:devices:*",
        "i:deviceprofiles:*",
        "x:scenes:*",
        "x:devices:$"
    ],
    "redirectUris": []
}
```

Installing or updating my SmartApp still results in the same 403 error. I tried several remedies to correct it: uninstalling/installing the SmartApp again in the app, restarting my server several times, including and excluding the `permissions()` directive (my understanding from you, though, is that `permissions()` was unnecessary for my use case).

Is it possible that the `SECURITY_ARM_STATE` event is still not fully supported?

Aside: including the `permissions()` directive with any or all permissions still causes an error when updating the SmartApp. Both of the following will result in a "Network or server error occurred. Try again later".
```
new SmartApp()
	// ...
	.permissions([
		'r:security:locations:*:armstate'
	])
```

```
new SmartApp()
	// ...
	.permissions([
		'r:devices:*',
		'x:devices:*',
		'r:security:locations:*:armstate'
		'i:deviceprofiles:*',
		'r:locations:*',
		'r:devices:*',
		'w:devices:*',
		'x:devices:*',
		'r:scenes:*',
		'x:scenes:*',
	])
```

**Comment by bflorian** [comment]
The security event is still supported. You need to use a `permissions()` directive whenever your app needs a scope that is not represented by a configuration setting. So for example, if you app has settings for selecting motion sensors and lights, the app itself needs `r:devices` and `x:devices` permissions, but you don't need to specify them in the `permissions()` directive.

Security arm state, on the other hand, doesn't have a configuration setting, so you need to specify it explicitly. You also need to use the `appId()` directive when you do that, so something like this should do it:
```
new SmartApp()
	// ...
	.appId('1010c506-ae05-45e3-9589-b7f8c0112eba') // your appId
	.permissions([
		'r:security:locations:*:armstate'
	])
```

**Comment by mattstrom** [comment]
I figured out why `permissions()` wasn't working for me. I hacw organized my SmartApp code within a class, and so I needed to append `.bind(this)` when wiring up my event handlers.

It's still not immediately clear to me why binding functions avoids the 403 error because the error only occurs when I use the `permissions()` directive with a non-empty array. But I have a solution now. Thank you for your help.
